### PR TITLE
support parallel builds

### LINKED ISSUE #34: spelling directive doesn't work with parallel builds
So I've tried adding a spelling directive with a list of words to the top of one RST file and it doesn't seem to work at all.
Any ideas on where do I start debugging?

I've looked into the source code and it seems it should do some logging but I don't see anything in the console output...

### LINKED ISSUE #55: Latest release erroring on sphinx-build with: "the sphinxcontrib.spelling extension is not safe for parallel reading"
```
Traceback (most recent call last):
  File "/home/travis/build/pyca/cryptography/.tox/docs/lib/python3.8/site-packages/sphinx/cmd/build.py", line 280, in build_main
    app.build(args.force_all, filenames)
  File "/home/travis/build/pyca/cryptography/.tox/docs/lib/python3.8/site-packages/sphinx/application.py", line 348, in build
    self.builder.build_update()
  File "/home/travis/build/pyca/cryptography/.tox/docs/lib/python3.8/site-packages/sphinx/builders/__init__.py", line 297, in build_update
    self.build(to_build,
  File "/home/travis/build/pyca/cryptography/.tox/docs/lib/python3.8/site-packages/sphinx/builders/__init__.py", line 311, in build
    updated_docnames = set(self.read())
  File "/opt/python/3.8.1/lib/python3.8/contextlib.py", line 120, in __exit__
    next(self.gen)
  File "/home/travis/build/pyca/cryptography/.tox/docs/lib/python3.8/site-packages/sphinx/util/logging.py", line 213, in pending_warnings
    memhandler.flushTo(logger)
  File "/home/travis/build/pyca/cryptography/.tox/docs/lib/python3.8/site-packages/sphinx/util/logging.py", line 178, in flushTo
    logger.handle(record)
  File "/opt/python/3.8.1/lib/python3.8/logging/__init__.py", line 1587, in handle
    self.callHandlers(record)
  File "/opt/python/3.8.1/lib/python3.8/logging/__init__.py", line 1649, in callHandlers
    hdlr.handle(record)
  File "/opt/python/3.8.1/lib/python3.8/logging/__init__.py", line 946, in handle
    rv = self.filter(record)
  File "/opt/python/3.8.1/lib/python3.8/logging/__init__.py", line 807, in filter
    result = f.filter(record)
  File "/home/travis/build/pyca/cryptography/.tox/docs/lib/python3.8/site-packages/sphinx/util/logging.py", line 417, in filter
    raise SphinxWarning(message)
sphinx.errors.SphinxWarning: the sphinxcontrib.spelling extension is not safe for parallel reading
```

This is happening on a build that's not even using the sphinx builder:

https://travis-ci.org/github/pyca/cryptography/jobs/701818891#L454

### PR DESCRIPTION
**Pull request by dhellmann**
Fixes #55
Fixes #34

### CODE REVIEW

File: sphinxcontrib/spelling/__init__.py [hunk 1]
```diff
@@ -43,4 +47,4 @@ def setup(app):
     app.add_config_value('spelling_ignore_importable_modules', True, 'env')
     # Add any user-defined filter classes
     app.add_config_value('spelling_filters', [], 'env')
-    return {"parallel_read_safe": False}
+    return {"parallel_read_safe": True}
```
Comment (webknjaz) [inline sphinxcontrib/spelling/__init__.py:50]: It'd be useful to explicitly declare the parallel write support level and the version:
```suggestion
    return {
        'parallel_read_safe': True,
        'parallel_write_safe': True,
        'version': __version__,
    }
```

(import the version somewhere on top of read via `importlib.metadata`/`pkg_resources`)

Comment (dhellmann) [reply sphinxcontrib/spelling/__init__.py]: Good idea. Done.

File: sphinxcontrib/spelling/asset.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+    def clear_doc(self, app, env, docname) -> None:
+        if docname in getattr(env, 'spelling_document_words', {}):
+            del env.spelling_document_words[docname]
+
+    def merge_other(self, app, env, docnames, other):
+        other_words = getattr(other, 'spelling_document_words', {})
+        if not hasattr(env, 'spelling_document_words'):
+            env.spelling_document_words = collections.defaultdict(list)
+        env.spelling_document_words.update(other_words)
```
Comment (webknjaz) [inline sphinxcontrib/spelling/asset.py:26]: How about using more EAFP
```suggestion
import contextlib

from sphinx.environment.collectors import EnvironmentCollector
from sphinx.util import logging

logger = logging.getLogger(__name__)


class SpellingCollector(EnvironmentCollector):

    def clear_doc(self, app, env, docname) -> None:
        with contextlib.suppress(AttributeError, KeyError):
            del env.spelling_document_words[docname]

    def merge_other(self, app, env, docnames, other):
        try:
            other_words = other.spelling_document_words
        except AttributeError:
            other_words = {}

        if not hasattr(env, 'spelling_document_words'):
            env.spelling_document_words = collections.defaultdict(list)
        env.spelling_document_words.update(other_words)
```

File: sphinxcontrib/spelling/checker.py [hunk 3]
```diff
@@ -5,8 +5,12 @@
...
 """
 
-import enchant
-from enchant.tokenize import get_tokenizer
+try:
+    import enchant
+    from enchant.tokenize import get_tokenizer
+    have_enchant = True
+except ImportError:
+    have_enchant = False
```
Comment (webknjaz) [inline sphinxcontrib/spelling/checker.py:13]: Usually, it's better to have less things (ideally just one instruction) in the `try` section:
```suggestion
except ImportError as imp_exc:
    enchant_import_error = imp_exc
else:
    enchant_import_error = None
```
Also, preserving the import error may help convey more details later.

File: sphinxcontrib/spelling/checker.py [hunk 4]
```diff
@@ -18,6 +22,10 @@ class SpellingChecker(object):
 
     def __init__(self, lang, suggest, word_list_filename,
                  tokenizer_lang='en_US', filters=None, context_line=False):
+        if not have_enchant:
+            raise RuntimeError(
+                'Cannot instantiate SpellingChecker '
+                'without PyEnchant installed')
```
Comment (webknjaz) [inline sphinxcontrib/spelling/checker.py:28]: Here's how you can chain the import error to show more details about the exception that caused it in the first place:
```suggestion
        if enchant_import_error is not None:
            raise RuntimeError(
                'Cannot instantiate SpellingChecker '
                'without PyEnchant installed',
            ) from enchant_import_error
```
(together with https://github.com/sphinx-contrib/spelling/pull/59/files#r460555742)

File: sphinxcontrib/spelling/__init__.py [hunk 5]
```diff
@@ -43,4 +55,7 @@ def setup(app):
     app.add_config_value('spelling_ignore_importable_modules', True, 'env')
     # Add any user-defined filter classes
     app.add_config_value('spelling_filters', [], 'env')
-    return {"parallel_read_safe": False}
+    return {
+        "parallel_read_safe": True,
```
Comment (webknjaz) [inline sphinxcontrib/spelling/__init__.py:59]: Is it also write-safe?

File: sphinxcontrib/spelling/asset.py [hunk 6]
```diff
@@ -0,0 +1,29 @@
+# encoding: utf-8
+#
+# Copyright (c) 2020 Doug Hellmann.  All rights reserved.
+#
+"""Asset collector for additional spelling terms
+"""
```
Comment (webknjaz) [inline sphinxcontrib/spelling/asset.py:6]: PEP257 requires a trailing period here
```suggestion
"""Asset collector for additional spelling terms."""
```

File: docs/source/install.rst [hunk 7]
```diff
@@ -1,24 +1,22 @@
...
+   .. code-block:: python
 
-1. Add ``'sphinxcontrib.spelling'`` to the ``extensions`` list in ``conf.py``.
+      extensions = [ 'sphinxcontrib.spelling' ]
 
-  ::
+3. Then pass ``"spelling"`` as the builder argument to ``sphinx-build``.
 
-    extensions = [ 'sphinxcontrib.spelling' ]
+   .. code-block:: shell
```
Comment (webknjaz) [inline docs/source/install.rst:20]: It should be `shell-session` or `console` per https://pygments.org/docs/lexers/#pygments.lexers.shell.BashSessionLexer
```suggestion
   .. code-block:: shell-session
```

File: sphinxcontrib/spelling/builder.py [hunk 8]
```diff
@@ -33,14 +37,18 @@ class SpellingBuilder(Builder):
     name = 'spelling'
 
     def init(self):
+        if not have_enchant:
+            raise RuntimeError(
+                'Cannot initialize spelling builder '
+                'without PyEnchant installed')
```
Comment (webknjaz) [inline sphinxcontrib/spelling/builder.py:43]: I suggest applying https://github.com/sphinx-contrib/spelling/pull/59/files#r460555871 here too

**Comment by webknjaz** [comment]
@dhellmann you should probably add "Fixes #34" to the PR description too.

File: sphinxcontrib/spelling/__init__.py (hunk 5 above)
Comment (dhellmann) [reply sphinxcontrib/spelling/__init__.py:59]: It could be made to be, but it probably isn't since it logs all output to a single file for review.

**Comment by dhellmann** [comment]
Thanks for the comprehensive review @webknjaz. I have incorporated all of your suggestions, except related to parallel writes. I'll tackle that in a separate PR, after I figure out what would be needed.

File: sphinxcontrib/spelling/__init__.py (hunk 5 above)
Comment (webknjaz) [reply sphinxcontrib/spelling/__init__.py:59]: @dhellmann I must warn you that **it is `True` by default**: https://www.sphinx-doc.org/en/master/extdev/index.html#extension-metadata.

So by not specifying it you basically announce that it's write-safe.

**Comment by webknjaz** [comment]
@dhellmann great! As for write-safe, I have a feeling that you expect it to be False if not specified, but it's the opposite: https://github.com/sphinx-contrib/spelling/pull/59#discussion_r460582231.

File: sphinxcontrib/spelling/__init__.py (hunk 5 above)
Comment (dhellmann) [reply sphinxcontrib/spelling/__init__.py:59]: Done

File: sphinxcontrib/spelling/builder.py [hunk 9]
```diff
@@ -9,14 +9,21 @@
 import importlib
 import io
 import os
+import os.path
```
Comment (webknjaz) [inline sphinxcontrib/spelling/builder.py:12]: given the previous line, this seems redundant
```suggestion
```

Comment (dhellmann) [reply sphinxcontrib/spelling/builder.py]: Done

**Comment by webknjaz** [comment]
:tada: https://pypi.org/project/sphinxcontrib-spelling/5.2.0/

**Pull request merged**